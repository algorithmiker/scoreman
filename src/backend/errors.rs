use std::{
    cmp::{max, min},
    fmt::{self, Display},
    ops::RangeInclusive,
};

use yansi::Paint;

use crate::collect_parse_error;

#[derive(Clone, Debug)]
pub enum ErrorLocation {
    NoLocation,
    LineOnly(usize),
    LineAndMeasure(usize, usize),
    LineAndCharIdx(usize, usize),
}
/// Produced [diagnostics.len()] diagnostics and one error.
/// Diagnostics:
///  - [location]
///    [severity] [message]
///
/// [short]
/// [location]
/// [long]
///
#[derive(Debug)]
pub struct BackendError<'a> {
    pub main_location: ErrorLocation,
    pub relevant_lines: RangeInclusive<usize>,
    pub kind: BackendErrorKind<'a>,
    pub diagnostics: Vec<Diagnostic>,
}

#[derive(Debug)]
pub enum BackendErrorKind<'a> {
    IOError(std::io::Error),
    FmtError(std::fmt::Error),
    EmptyScore,
    /// string name and fret
    NoSuchFret(char, u16),
    TickMismatch(char, char, usize, usize),
    /// string name where tick is multichar, string name here, tick idx, and the found invalid fret
    BadMulticharTick {
        /// string and fret
        multichar: (char, u16),
        /// string and fret
        invalid: (char, u16),
        tick_idx: usize,
    },
    InvalidCommentSyntax(String),
    InvalidPartlineSyntax(String),
    ParseError(nom::Err<nom::error::VerboseError<&'a str>>),
}

impl<'a> BackendErrorKind<'a> {
    pub fn desc(&self) -> (String, String) {
        match self {
            BackendErrorKind::IOError(x) => {
                ("Cannot write to file".into(), format!("IO Error:\n{x}"))
            }

            BackendErrorKind::FmtError(x) => {
                ("Cannot write to internal buffer".into(), format!("Format error:\n{x}"))
            }
            BackendErrorKind::EmptyScore => ("Empty score".into(), String::new()),
            BackendErrorKind::NoSuchFret(string_name, fret) => (
                "No such fret".into(),
                format!("Failed to get note for fret {fret} on string {string_name}"),
            ),

            BackendErrorKind::TickMismatch(string_before, string_after,ticks_before, ticks_after) => ("Tick mismatch".into(),
format!("The muxml2 backend relies on the fact that there are the same number of ticks (frets/rests) on every line (string) of a measure in the tab. This is not true for this tab.
The measure has {ticks_before} ticks on string {string_before} and {ticks_after} ticks on string {string_after}.

Tip: If you get a lot of errors like this, consider using the muxml1 backend.")
            ),
            BackendErrorKind::BadMulticharTick { multichar : (multichar_string,multichar_fret), invalid: (invalid_string,invalid_fret), tick_idx } =>
            (
                "Invalid multichar tick".into(), 
                format!(
"Tick {} has a multi-char fret ({multichar_fret}) on string {multichar_string}, but on the same tick there is an invalid fret {invalid_fret} on string {invalid_string}", tick_idx+1)
            ),
            BackendErrorKind::InvalidCommentSyntax(rem) => ("Invalid comment syntax".into(), format!("Got remaining content: `{rem}`")),
            BackendErrorKind::InvalidPartlineSyntax(rem) => ("Invalid partline syntax".into(), format!("Got remaining content: `{rem}`")),
            BackendErrorKind::ParseError(x) => ("Invalid syntax".into(), collect_parse_error(x)),

        }
    }
}
#[derive(Debug, Clone)]
pub enum DiagnosticSeverity {
    Info,
    Warning,
}
impl Display for DiagnosticSeverity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                DiagnosticSeverity::Info => "Info".blue(),
                DiagnosticSeverity::Warning => "Warning".yellow(),
            }
        )
    }
}
#[derive(Debug, Clone)]
pub struct Diagnostic {
    pub location: ErrorLocation,
    pub message: String,
    pub severity: DiagnosticSeverity,
}

impl Diagnostic {
    pub fn info(location: ErrorLocation, message: String) -> Self {
        Self {
            location,
            message,
            severity: DiagnosticSeverity::Info,
        }
    }
    pub fn warn(location: ErrorLocation, message: String) -> Self {
        Self {
            location,
            message,
            severity: DiagnosticSeverity::Warning,
        }
    }
}

impl<'a> BackendError<'a> {
    pub fn from_io_error(x: std::io::Error, diagnostics: Vec<Diagnostic>) -> Self {
        BackendError {
            main_location: ErrorLocation::NoLocation,
            kind: BackendErrorKind::IOError(x),
            relevant_lines: 0..=0,
            diagnostics,
        }
    }
    pub fn from_fmt_error(x: std::fmt::Error, diagnostics: Vec<Diagnostic>) -> Self {
        BackendError {
            main_location: ErrorLocation::NoLocation,
            //short: "Cannot write to internal buffer".to_string(),
            //long: format!("Format error:\n{x}"),
            diagnostics,
            relevant_lines: 0..=0,
            kind: BackendErrorKind::FmtError(x),
        }
    }

    pub fn empty_score_err(diagnostics: Vec<Diagnostic>) -> Self {
        BackendError {
            main_location: ErrorLocation::NoLocation,
            diagnostics,
            kind: BackendErrorKind::EmptyScore,
            relevant_lines: 0..=0,
        }
    }

    pub fn no_such_fret(
        location_a: usize,
        location_b: usize,
        string_name: char,
        fret: u16,
        diagnostics: Vec<Diagnostic>,
    ) -> Self {
        Self {
            main_location: ErrorLocation::LineAndMeasure(location_a, location_b),
            diagnostics,
            kind: BackendErrorKind::NoSuchFret(string_name, fret),
            relevant_lines: location_a..=location_a,
        }
    }
}

impl ErrorLocation {
    pub fn get_line_idx(&self) -> Option<usize> {
        match self {
            ErrorLocation::NoLocation => None,
            ErrorLocation::LineOnly(x) => Some(*x),
            ErrorLocation::LineAndMeasure(x, _) => Some(*x),
            ErrorLocation::LineAndCharIdx(x, _) => Some(*x),
        }
    }
    pub fn write_location_explainer(&self, f: &mut impl std::fmt::Write) {
        match self {
            ErrorLocation::NoLocation => (),
            ErrorLocation::LineOnly(line_idx) => {
                let line_num = line_idx + 1;
                writeln!(f, "{} in line {line_num}:", "Where:".bold(),).unwrap();
            }
            ErrorLocation::LineAndMeasure(line_idx, measure_idx) => {
                let (line_num, measure_num) = (line_idx + 1, measure_idx + 1);
                writeln!(
                    f,
                    "{} Measure {measure_num} in line {line_num}:",
                    "Where:".bold()
                )
                .unwrap();
            }
            ErrorLocation::LineAndCharIdx(line_idx, char_idx) => writeln!(
                f,
                "{} line {} char {}",
                "Where:".bold(),
                line_idx + 1,
                char_idx + 1,
            )
            .unwrap(),
        }
    }
}

pub const ERROR_CONTEXT: usize = 3;

/// When reporting an error with `relevant_lines`, we want to show some context of +=, [ERROR_CONTEXT]
/// lines that are not neccessarily relevant
/// This is not always possible because of line bounds (cannot show 3 lines before line 0)
/// This function handles that.
pub fn extend_error_range(range: &RangeInclusive<usize>, line_cnt: usize) -> RangeInclusive<usize> {
    let start = max(*range.start(), ERROR_CONTEXT) - ERROR_CONTEXT;
    let end = min(line_cnt, range.end() + ERROR_CONTEXT);

    start..=end
}

pub fn get_digit_cnt(num: usize) -> u32 {
    num.checked_ilog10().unwrap_or(0) + 1
}
