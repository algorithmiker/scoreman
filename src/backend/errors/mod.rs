use std::{
    cmp::{max, min},
    ops::RangeInclusive,
};

pub mod backend_error;
pub mod backend_error_kind;
pub mod diagnostic;
pub mod diagnostic_kind;
pub mod error_location;

pub const ERROR_CONTEXT: usize = 3;

/// When reporting an error with `relevant_lines`, we want to show some context of +=, [ERROR_CONTEXT]
/// lines that are not neccessarily relevant
/// This is not always possible because of line bounds (cannot show 3 lines before line 0)
/// This function handles that.
pub fn extend_error_range(range: &RangeInclusive<usize>, line_cnt: usize) -> RangeInclusive<usize> {
    let start = max(*range.start(), ERROR_CONTEXT) - ERROR_CONTEXT;
    let end = min(line_cnt - 1, range.end() + ERROR_CONTEXT);

    start..=end
}
