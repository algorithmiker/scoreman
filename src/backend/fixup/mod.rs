use std::time::{Duration, Instant};

use clap::ValueEnum;

use super::BackendResult;
use crate::{
    backend::{
        errors::{
            backend_error::BackendError, backend_error_kind::BackendErrorKind,
            diagnostic::Diagnostic, diagnostic_kind::DiagnosticKind, error_location::ErrorLocation,
        },
        Backend,
    },
    parser::parser::parse,
    time, traceln,
};

pub struct FixupBackend();
#[derive(ValueEnum, Clone)]
pub enum FixupDumpOptions {
    TickStream,
    PrettyTracks,
}
#[derive(Clone)]
pub struct FixupBackendSettings {
    pub dump: Option<FixupDumpOptions>,
}

struct LocationTracker {
    pub data: [ErrorLocation; 5],
    pub push_cnt: u32,
}

impl LocationTracker {
    pub fn new() -> Self {
        Self { data: [const { ErrorLocation::NoLocation }; 5], push_cnt: 0 }
    }
    pub fn add(&mut self, l: ErrorLocation) {
        traceln!("location_tracker:: before add: {:?}", self.data);
        // shift left
        for i in 0..4 {
            self.data[i] = self.data[i + 1].clone();
        }
        self.data[4] = l;
        traceln!("location_tracker:: after add: {:?}", self.data);
        self.push_cnt += 1;
    }
    pub fn is_same(&self) -> bool {
        self.push_cnt > 4 && self.data.windows(2).all(|a| a[0] == a[1])
    }
}
impl Backend for FixupBackend {
    type BackendSettings = FixupBackendSettings;

    fn process<Out: std::io::Write>(
        parser_input: &[String], out: &mut Out, settings: Self::BackendSettings,
    ) -> BackendResult {
        if let Some(dump) = settings.dump {
            let (parse_time, parsed) = time(|| parse(parser_input));
            let mut r = BackendResult::new(vec![], None, Some(parse_time), None);
            match dump {
                FixupDumpOptions::TickStream => writeln!(out, "{:?}", parsed.tick_stream).unwrap(),
                FixupDumpOptions::PrettyTracks => {
                    writeln!(out, "{}", parsed.dump_tracks()).unwrap()
                }
            }
            r.err = parsed.error;
            return r;
        }
        let mut diagnostics = vec![];
        // TODO: figure out a way not to clone these
        let mut parser_input = parser_input.to_owned();
        let mut parse_time = Duration::from_secs(0);
        let fixup_start = Instant::now();
        let mut location_tracker = LocationTracker::new();
        loop {
            let parse_start = Instant::now();
            let parsed = parse(&parser_input);
            parse_time = parse_start.elapsed();
            match &parsed.error {
                None => break,
                Some(err) => {
                    location_tracker.add(err.main_location.clone());
                    if location_tracker.is_same() {
                        let lines =
                            err.main_location.get_line_idx().map(|x| x..=x).unwrap_or(0..=0);
                        return BackendResult::new(
                            diagnostics,
                            Some(BackendError::fixup_failed(err.main_location.clone(), lines)),
                            Some(parse_time),
                            Some(fixup_start.elapsed()),
                        );
                    }
                    match err.kind {
                        BackendErrorKind::IOError(_) | BackendErrorKind::FmtError(_) => {
                            return BackendResult::new(
                                diagnostics,
                                parsed.error,
                                Some(parse_time),
                                None,
                            );
                        }
                        BackendErrorKind::BendOnInvalid => {} // todo: bendOnInvalid fixup: remove the bend
                        BackendErrorKind::InvalidStringName => {}
                        BackendErrorKind::EmptyScore => {}
                        BackendErrorKind::BothSlotsMultiChar => {} // todo: fix BothSlotsMultichar errors
                        BackendErrorKind::FretTooLarge => {} // todo: fix FretTooLarge errors (add
                        // space between)
                        BackendErrorKind::MultiBothSlotsFilled => {
                            let Some((line_idx, char_idx)) = err
                                .main_location
                                .get_line_idx()
                                .zip(err.main_location.get_char_idx())
                            else {
                                continue;
                            };
                            // we just replace both with a rest to be sure
                            // todo: use a better strategy by checking the actual ticks
                            parser_input[line_idx].replace_range(char_idx..char_idx + 1, "-");
                            parser_input[line_idx].replace_range(char_idx + 1..char_idx + 2, "-");
                            diagnostics.push(Diagnostic::info(
                                err.main_location.clone(),
                                DiagnosticKind::FormatReplacedInvalid,
                            ));
                        }
                        BackendErrorKind::NoClosingBarline => {
                            let l_idx = err.main_location.get_line_idx().unwrap();
                            let line = &mut parser_input[l_idx];
                            line.truncate(line.trim_end().len());
                            line.push('|');
                            diagnostics.push(Diagnostic::info(
                                err.main_location.clone(),
                                DiagnosticKind::FormatAddedBarline,
                            ))
                        }
                        BackendErrorKind::FixupFailed => unreachable!(),
                        BackendErrorKind::Parse3InvalidCharacter(_) => {
                            let Some((line_idx, char_idx)) = err
                                .main_location
                                .get_line_idx()
                                .zip(err.main_location.get_char_idx())
                            else {
                                continue;
                            };
                            parser_input[line_idx].replace_range(char_idx..char_idx + 1, "-");
                            diagnostics.push(Diagnostic::info(
                                err.main_location.clone(),
                                DiagnosticKind::FormatReplacedInvalid,
                            ))
                        }
                    };
                }
            }
        }
        let gen_time = fixup_start.elapsed();
        let maybe_io_err =
            out.write_all(parser_input.join("\n").as_ref()).map_err(BackendError::from).err();
        BackendResult::new(diagnostics, maybe_io_err, Some(parse_time), Some(gen_time))
    }
}
