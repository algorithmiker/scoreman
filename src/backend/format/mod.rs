use super::{Backend, BackendError, Diagnostic};
use crate::parser::Section;

pub struct FormatBackend();

impl Backend for FormatBackend {
    type BackendSettings = ();

    fn process<Out: std::io::Write>(
        score: crate::parser::Score,
        out: &mut Out,
        _settings: Self::BackendSettings,
    ) -> Result<Vec<Diagnostic>, BackendError> {
        let diagnostics = vec![];
        let mut formatted = String::new();
        let mut measure_cnt = 0;
        for section in score.0 {
            match section {
                Section::Part { part, .. } => {
                    let measures_in_part = part[0].staffs.len();
                    for measure_idx in 0..measures_in_part {
                        formatted += &format!("// SYS: Measure {}\n", measure_cnt + 1);
                        for line in &part {
                            formatted.push(line.string_name);
                            formatted.push('|');
                            formatted += &line.staffs[measure_idx].print_pretty_string();
                            formatted.push('|');
                            formatted.push('\n');
                        }
                        formatted.push('\n');
                        measure_cnt += 1;
                    }
                }
                Section::Comment(x) => {
                    // SYS-comments were generated during a previous format run, so don't include
                    // them
                    if !x.trim_start().starts_with("SYS:") {
                        formatted += "//";
                        formatted += &x;
                        formatted.push('\n');
                    }
                }
            }
        }

        if let Err(x) = out.write_all(formatted.as_bytes()) {
            return Err(BackendError::from_io_error(x, diagnostics));
        }

        Ok(diagnostics)
    }
}
