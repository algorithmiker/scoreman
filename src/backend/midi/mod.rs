use std::collections::HashMap;

use midly::{
    num::{u28, u7},
    Format, Header, MetaMessage, MidiMessage, Smf, TrackEvent, TrackEventKind,
};

use crate::parser::{parser2::Parse2Result, TabElement::*};

use super::{
    errors::{backend_error::BackendError, diagnostic::Diagnostic},
    Backend,
};

const BPM: u32 = 80;
const MINUTE_IN_MICROSECONDS: u32 = 60 * 1000;
const LENGTH_OF_QUARTER: u32 = MINUTE_IN_MICROSECONDS / BPM;
const LENGTH_OF_EIGHT: u32 = LENGTH_OF_QUARTER / 2;

pub struct MidiBackend();
impl Backend for MidiBackend {
    type BackendSettings = ();

    fn process<Out: std::io::Write>(
        parse_result: Parse2Result,
        out: &mut Out,
        _settings: Self::BackendSettings,
    ) -> Result<Vec<Diagnostic>, BackendError> {
        let diagnostics = vec![];
        // TODO: the parser now gives us things like tick count, can probably preallocate based on
        // that
        let mut midi_tracks = convert_to_midi(parse_result);
        let mut tracks = vec![vec![
            TrackEvent {
                delta: 0.into(),
                kind: TrackEventKind::Meta(MetaMessage::TimeSignature(4, 4, 1, 8)),
            },
            TrackEvent {
                delta: 0.into(),
                kind: TrackEventKind::Meta(MetaMessage::Tempo(LENGTH_OF_QUARTER.into())),
            },
            TrackEvent {
                delta: 0.into(),
                kind: TrackEventKind::Meta(MetaMessage::EndOfTrack),
            },
        ]];
        tracks.append(&mut midi_tracks);
        let smf = Smf {
            header: Header::new(Format::Parallel, midly::Timing::Metrical(4.into())),
            tracks,
        };
        if let Err(x) = smf.write_std(out) {
            return Err(BackendError::from_io_error(x, diagnostics));
        }
        Ok(diagnostics)
    }
}

fn convert_to_midi(parse_result: Parse2Result) -> Vec<Vec<TrackEvent<'static>>> {
    let mut string_freq = HashMap::new();
    string_freq.insert('E', 52);
    string_freq.insert('A', 57);
    string_freq.insert('D', 62);
    string_freq.insert('G', 67);
    string_freq.insert('B', 71);
    string_freq.insert('d', 74);
    string_freq.insert('e', 76);
    let mut tracks: Vec<Vec<TrackEvent>> = vec![Vec::new(); 6];

    #[allow(clippy::needless_range_loop)]
    for i in 0..6 {
        let string_name = parse_result.string_names[i];
        let raw_track = &parse_result.strings[i];
        let mut delta_carry: u32 = 0;
        for raw_tick in raw_track {
            match raw_tick.element {
                Fret(fret) => {
                    let pitch = fret + string_freq[&string_name];
                    let (note_on, note_off) = gen_note_events(pitch.into(), delta_carry.into());
                    delta_carry = 0;
                    tracks[i].push(note_on);
                    tracks[i].push(note_off);
                }
                Rest => delta_carry += LENGTH_OF_EIGHT,
                // dead notes are purely cosmetic in this implementation
                DeadNote => (),
            }
        }
        tracks[i].push(TrackEvent {
            delta: 0.into(),
            kind: TrackEventKind::Meta(MetaMessage::EndOfTrack),
        });
    }
    tracks
}
fn gen_note_events<'a>(key: u7, initial_delta: u28) -> (TrackEvent<'a>, TrackEvent<'a>) {
    let note_on = TrackEvent {
        delta: initial_delta,
        kind: TrackEventKind::Midi {
            channel: 0.into(),
            message: MidiMessage::NoteOn {
                key,
                vel: 100.into(),
            },
        },
    };

    let note_off = TrackEvent {
        delta: LENGTH_OF_EIGHT.into(),
        kind: TrackEventKind::Midi {
            channel: 0.into(),
            message: MidiMessage::NoteOff {
                key,
                vel: 100.into(),
            },
        },
    };
    (note_on, note_off)
}
