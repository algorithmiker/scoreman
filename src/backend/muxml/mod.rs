use crate::{
    backend::errors::{diagnostic_kind::DiagnosticKind, error_location::ErrorLocation},
    parser::{
        Measure, Score,
        TabElement::{self, Fret},
    },
};

use super::{muxml2::fretboard::get_fretboard_note2, Backend, BackendError, Diagnostic};

pub struct MuxmlBackend();
impl Backend for MuxmlBackend {
    type BackendSettings = ();

    fn process<Out: std::io::Write>(
        score: Score,
        out: &mut Out,
        _settings: Self::BackendSettings,
    ) -> Result<Vec<Diagnostic>, BackendError> {
        use ErrorLocation::*;
        let mut diagnostics = vec![Diagnostic::warn(NoLocation, DiagnosticKind::Muxml1IsBad)];
        let (raw_tracks, _) = score.gen_raw_tracks()?;
        let (xml_out, mut xml_diagnostics) = raw_tracks_to_xml(raw_tracks)?;
        diagnostics.append(&mut xml_diagnostics);
        diagnostics.push(Diagnostic::info(
            NoLocation,
            DiagnosticKind::Muxml1SeperateTracks,
        ));
        if let Err(x) = out.write_all(xml_out.as_bytes()) {
            return Err(BackendError::from_io_error(x, diagnostics));
        }
        Ok(diagnostics)
    }
}

fn raw_tracks_to_xml<'a>(
    raw_tracks: ([char; 6], [Vec<Measure>; 6]),
) -> Result<(String, Vec<Diagnostic>), BackendError<'a>> {
    let mut parts_xml = String::new();
    let diagnostics = vec![];
    for i in 0..6 {
        let part = &raw_tracks.1[i];
        let mut measures_xml = String::new();
        for (measure_idx, measure) in part.iter().enumerate() {
            let mut notes_xml = String::new();
            for raw_tick in &measure.content {
                match raw_tick.element {
                    Fret(fret) => {
                        let x = get_fretboard_note2(raw_tracks.0[i], fret)?;
                        x.write_muxml(&mut notes_xml, false).unwrap();
                    }
                    TabElement::DeadNote => {
                        let mut x = get_fretboard_note2(raw_tracks.0[i], 0)?;
                        x.dead = true;
                        x.write_muxml(&mut notes_xml, false).unwrap();
                    }
                    TabElement::Rest => notes_xml.push_str(&muxml_rest("eighth", 1)),
                }
            }
            //println!("[D]: finished {measure:?}");
            measures_xml.push_str(&muxml_measure(
                measure_idx as u32,
                measure.content.len(),
                &notes_xml,
            ));
        }
        // musescore numbers p1 to p6, so we do that too for nicer diffs
        parts_xml.push_str(&muxml_part(i as u32 + 1, &measures_xml));
    }

    Ok((muxml_document(&parts_xml), diagnostics))
}

fn muxml_rest(r#type: &str, duration: u8) -> String {
    format!(
        r#"
<note>
  <rest measure="no"/>
  <duration>{duration}</duration>
  <voice>1</voice>
  <type>{type}</type>
</note>
"#
    )
}

fn muxml_measure(number: u32, note_count: usize, notes: &str) -> String {
    let (key, clef) = if number == 0 {
        (
            r#"<key>
  <fifths>0</fifths>
</key>"#,
            r#"<clef>
  <sign>G</sign>
  <line>2</line>
</clef>
"#,
        )
    } else {
        ("", "")
    };

    format!(
        r#"
<measure number="{number}">
  <attributes>
    <divisions>2</divisions>
    {key}
    <time>
      <beats>{note_count}</beats>
      <beat-type>8</beat-type>
    </time>
    {clef}
  </attributes>
  {notes}
</measure>
"#,
    )
}

fn muxml_part(number: u32, measures: &str) -> String {
    format!(
        r#"
<part id="P{number}">
  {measures}
</part>
    "#
    )
}

fn muxml_document(parts: &str) -> String {
    format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE score-partwise PUBLIC "-//Recordare//DTD MusicXML 4.0 Partwise//EN" "http://www.musicxml.org/dtds/partwise.dtd">
<score-partwise version="4.0">
  <identification>
    <encoding>
      <software>guitar_tab</software>
      <supports element="accidental" type="yes"/>
      <supports element="beam" type="yes"/>
      <supports element="print" attribute="new-page" type="no"/>
      <supports element="print" attribute="new-system" type="no"/>
      <supports element="stem" type="yes"/>
    </encoding>
  </identification>
  <part-list>
    <score-part id="P1">
      <part-name>Guitar1</part-name>
    </score-part>
    <score-part id="P2">
      <part-name>Guitar2</part-name>
    </score-part>
    <score-part id="P3">
      <part-name>Guitar3</part-name>
    </score-part>
    <score-part id="P4">
      <part-name>Guitar4</part-name>
    </score-part>
    <score-part id="P5">
      <part-name>Guitar5</part-name>
    </score-part>
    <score-part id="P6">
      <part-name>Guitar6</part-name>
    </score-part>
  </part-list>
  {parts}
</score-partwise>
"#
    )
}
