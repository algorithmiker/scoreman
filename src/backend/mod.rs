use errors::{backend_error::BackendError, diagnostic::Diagnostic};

use std::{fmt::Display, time::Duration};
pub mod errors;
pub mod fixup;
pub mod midi;
pub mod muxml;
pub mod muxml2;
pub struct BackendResult {
    pub diagnostics: Vec<Diagnostic>,
    pub err: Option<BackendError>,
    pub timing_parse: Option<Duration>,
    pub timing_gen: Option<Duration>,
}
impl BackendResult {
    pub fn new(
        diagnostics: Vec<Diagnostic>, err: Option<BackendError>, timing_parse: Option<Duration>,
        timing_gen: Option<Duration>,
    ) -> Self {
        Self { diagnostics, err, timing_parse, timing_gen }
    }
}
pub trait Backend {
    type BackendSettings;

    fn process<Out: std::io::Write>(
        input: &[String], out: &mut Out, settings: Self::BackendSettings,
    ) -> BackendResult;
}

/// Handles backend dispatch. Can be easily created from a string identifier
#[derive(Clone)]
pub enum BackendSelector {
    Midi(()),
    // Muxml(()),
    Muxml2(muxml2::settings::Settings),
    Fixup(fixup::FixupBackendSettings),
}

impl BackendSelector {
    // TODO: remove the lifetime here
    pub fn process<Out: std::io::Write>(self, input: &[String], out: &mut Out) -> BackendResult {
        match self {
            BackendSelector::Midi(settings) => midi::MidiBackend::process(input, out, settings),
            // BackendSelector::Muxml(settings) => muxml::MuxmlBackend::process(input, out, settings),
            BackendSelector::Muxml2(settings) => {
                muxml2::Muxml2Backend::process(input, out, settings)
            }
            BackendSelector::Fixup(settings) => fixup::FixupBackend::process(input, out, settings),
        }
    }
}

impl Display for BackendSelector {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                BackendSelector::Midi(_) => "midi",
                // BackendSelector::Muxml(_) => "muxml",
                BackendSelector::Muxml2(_) => "muxml",
                BackendSelector::Fixup(_) => "fixup",
            }
        )
    }
}
