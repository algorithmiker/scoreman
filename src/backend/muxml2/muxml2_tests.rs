use crate::backend::{
    muxml2::{settings::Settings, Muxml2Backend},
    Backend,
};

#[test]
fn test_muxml2() -> anyhow::Result<()> {
    let i1 = r#"
e|--------------|
B|-----------0-1|
G|-------1-2----|
D|--0-2-4-------|
A|-3------------|
E|--------------|
    "#;
    let mut out = vec![];
    let settings = Settings {
        remove_rest_between_notes: true,
        trim_measure: true,
        simplify_time_signature: true,
        bend_mode: crate::backend::muxml2::settings::Muxml2BendMode::EmulateBends,
    };
    Muxml2Backend::process(i1.lines(), &mut out, settings);
    insta::assert_snapshot!(String::from_utf8_lossy(&out));
    Ok(())
}
#[test]
fn test_muxml_bends() -> anyhow::Result<()> {
    let i1 = r#"
e|----|
B|----|
G|2b4-|
D|----|
A|----|
E|----|
    "#;
    let mut out = vec![];
    let settings = Settings {
        remove_rest_between_notes: true,
        trim_measure: true,
        simplify_time_signature: true,
        bend_mode: crate::backend::muxml2::settings::Muxml2BendMode::EmulateBends,
    };
    Muxml2Backend::process(i1.lines(), &mut out, settings);
    insta::assert_snapshot!(String::from_utf8_lossy(&out));
    Ok(())
}
