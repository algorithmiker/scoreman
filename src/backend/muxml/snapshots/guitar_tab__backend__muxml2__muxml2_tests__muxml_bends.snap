---
source: src/backend/muxml2/muxml2_tests.rs
expression: "String::from_utf8_lossy(&out)"
---
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE score-partwise PUBLIC "-//Recordare//DTD MusicXML 4.0 Partwise//EN" "http://www.musicxml.org/dtds/partwise.dtd">
<score-partwise version="4.0">
  <identification>
    <encoding>
      <software>guitar_tab</software>
      <supports element="accidental" type="yes"/>
      <supports element="beam" type="yes"/>
      <supports element="print" attribute="new-page" type="no"/>
      <supports element="print" attribute="new-system" type="no"/>
      <supports element="stem" type="yes"/>
    </encoding>
  </identification>
  <part-list>
    <score-part id="P1">
      <part-name>Guitar1</part-name>
    </score-part>
  </part-list>
  <part id="P1">
<measure number="0">
<attributes>
<divisions>2</divisions>
<key><fifths>0</fifths></key>
<time><beats>1</beats><beat-type>4</beat-type></time>
<clef><sign>G</sign><line>2</line></clef>
</attributes>
<note>
<pitch><step>A</step>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type>
<notations>
<slur type="start" number=1" />
</notations>
</note>
<note>
<pitch><step>A</step>
<alter>1</alter>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type>
<accidental>sharp</accidental>
<notations>
<slur type="start" number=1" />
</notations>
</note>
</measure>
</part>
</score-partwise>
